reindexer/reindex_worker/src/test/scala/uk/ac/wellcome/platform/reindex_worker/services/ReindexWorkerServiceTest.scala

```diff
@@ -0,0 +1,216 @@
+                      mock[AmazonCloudWatch],
+                      actorSystem)
+
+  override lazy val tableName = "reindex-worker-service-test"
+
+  override lazy val evidence: DynamoFormat[HybridRecord] =
+    DynamoFormat[HybridRecord]
+
+  private val enrichedDynamoFormat: DynamoFormat[HybridRecord] = Sourced
+    .toSourcedDynamoFormatWrapper[HybridRecord]
+    .enrichedDynamoFormat
+
+  val queueUrl = createQueueAndReturnUrl("reindex-worker-service-test-q")
+  val topicArn = createTopicAndReturnArn("reindex-worker-service-test-topic")
+
+  it("returns a successful Future if the reindex completes correctly") {
+    val reindexJob = ReindexJob(
+      shardId = "sierra/123",
+      desiredVersion = 6
+    )
+
+    val hybridRecord = HybridRecord(
+      version = 1,
+      sourceId = "sierra",
+      sourceName = "111",
+      s3key = "s3://reindexWST/example.json",
+      reindexShard = reindexJob.shardId,
+      reindexVersion = reindexJob.desiredVersion - 1
+    )
+
+    Scanamo.put(dynamoDbClient)(tableName)(hybridRecord)(enrichedDynamoFormat)
+
+    val expectedRecords = List(
+      hybridRecord.copy(version = hybridRecord.version + 1,
+                        reindexVersion = reindexJob.desiredVersion)
+    )
+
+    val invalidSqsMessage = SQSMessage(
+      subject = None,
+      body = toJson(reindexJob).get,
+      topic = "topic",
+      messageType = "message",
+      timestamp = "now"
+    )
+
+    val service = reindexWorkerService()
+
+    val future = service.processMessage(message = invalidSqsMessage)
+
+    whenReady(future) { _ =>
+      val actualRecords: List[HybridRecord] =
+        Scanamo.scan[HybridRecord](dynamoDbClient)(tableName).map(_.right.get)
+
+      actualRecords shouldBe expectedRecords
+    }
+  }
+
+  it(
+    "returns a failed Future if it cannot parse the SQS message as a ReindexJob") {
+    val sqsMessage = SQSMessage(
+      subject = None,
+      body = "<xml>What is JSON.</xl?>",
+      topic = "topic",
+      messageType = "message",
+      timestamp = "now"
+    )
+
+    val service = reindexWorkerService()
+
+    val future = service.processMessage(message = sqsMessage)
+
+    whenReady(future.failed) { result =>
+      result shouldBe a[GracefulFailureException]
+      result.getMessage should include(
+        "expected json value got < (line 1, column 1)")
+    }
+  }
+
+  it("returns a failed Future if the reindex job fails") {
+    val reindexJob = ReindexJob(
+      shardId = "sierra/333",
+      desiredVersion = 3
+    )
+
+    val sqsMessage = SQSMessage(
+      subject = None,
+      body = toJson(reindexJob).get,
+      topic = "topic",
+      messageType = "message",
+      timestamp = "now"
+    )
+
+    val service = reindexWorkerService(dynamoTableName = "does-not-exist")
+
+    val future = service.processMessage(message = sqsMessage)
+
+    whenReady(future.failed) { result =>
+      result shouldBe a[GracefulFailureException]
```

Comment:
Should this be a `GracefulFailureException`? a table not existing should trigger a terminal failure

Reply:
Yeah, this is a bit of a fudge – we catch everything and rewrap it as GracefulFailureException. 🙈